build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1029,6 +1029,54 @@ message ActionResult {
 
   // The details of the execution that originally produced this result.
   ExecutedActionMetadata execution_metadata = 9;
+
+  // Inlined [ConsoleOutputs][build.bazel.remote.execution.v2.ConsoleOutputs]
+  // containing a series of stdout and stderr outputs in order of write from
+  // the action. The server SHOULD NOT inline `console_outputs` unless
+  // requested by the client in the
+  // [GetActionResultRequest][build.bazel.remote.execution.v2.GetActionResultRequest]
+  // message. The server MAY omit inlining, even if requested, and MUST do so if inlining
+  // would cause the response to exceed message size limits.
+  ConsoleOutputs console_outputs = 10;
+
+  // The digest for a blob containing the
+  // [ConsoleOutputs][build.bazel.remote.execution.v2.ConsoleOutputs] for the action,
+  // which can be retrieved from the
+  // [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage].
+  Digest console_outputs_digest = 11;
+}
+
+// A sequence of console outputs in the order received for a remoted process.
+// This may contain the interleaved outputs from a process in the order
+// received during execution. If the service does not implement
+// interleaving, this may contain a summary of the stdout and stderr output
+// streams as single large writes in a server-defined order.
+message ConsoleOutputs
+{
+  // An output record for one or more outputs to stdout or stderr.
+  //
+  // Design note: For text coloring and other effects, ANSI codes may be
+  // embedded within these strings. Windows console API calls are not
+  // supported in this protocol, but a server implementation MAY choose to
+  // translate console API calls into ANSI codes, as more recent Windows 10
+  // releases included ANSI code support.
+  message ConsoleOutput {
+    enum Stream {
+      STDOUT = 0;
+      STDERR = 1;
+    }
+    
+    // The output stream to which the `Writes` were directed.
+    Stream stream = 1;
+    
+    // One or more written strings. There is an implicit operating system
```

Comment:
I don’t think we should add implicit newlines. The issue is that some POSIX applications mix stdout and stderr output on single lines. By using implicit newlines, there would be no way to express the output of those.

Though slightly contrived, be sure to read the POSIX spec of the `fuser` utility.

Reply:
I'm also in favour of explicit newlines. If an application that chooses to use LF instead of CRLF under Windows, the output would have to be concatenated into a single entry. Would one have to look into the platform fields of an action to find out if LF or CRLF should be used? How would outputs without trailing newline be handled with implicit newline?

Reply:
Alternatively, we could simply standardize on one convention. Even for UNIX it would make sense to use \r\n, as that is actually what’s emitted on the console. Userspace applications write \n, but termios converts it to \r\n in case ONLCR is enabled.

That said, I’m not 100% sure it’s safe to use pseudo-terminals instead of pipes on UNIX. That may cause applications to assume they need to be run interactively. Maybe the best strategy is to document that applications can both emit \n and \r\n, and that it’s up to the tool that displays the logs to seamlessly support both formats.

Reply:
Given the use case--sending back the "raw" console output as it would have appeared if the action had been run locally--I'm inclined toward's Ed's suggestion to make the API as "raw" as possible and let it be the responsibility of the client to interpret the output properly. This is similar reasoning to how we're handling absolute paths (which are OS-dependent) and relative paths (which are not, and therefore use a "generic" separator regardless of OS).

Reply:
For posterity: Tools often generate binary output and any post-processing of that output would break that. For example, gzip can write output to stdout, as can tar, as well as a variety of other CLI tools. Bazel does as well in some cases - this already broke Google-internal API definition that assumed that Bazel always outputs valid UTF-8. Post-processing stdout/stderr would be a very, very bad idea.